src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+    if (fakeDataBuf.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    uint8_t * fakeBlockData = fakeDataBuf->Start();
+
+    // Provide Block data and verify sender emits Block message
+    TransferSession::BlockData blockData;
+    blockData.Data   = fakeBlockData;
+    blockData.Length = maxBlockSize;
+    blockData.IsEof  = isEof;
+
+    err = sender.PrepareBlock(blockData);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle blockMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !blockMsgToSend.IsNull());
+
+    // Pass Block message to receiver and verify Block is received
+    err = receiver.HandleMessageReceived(blockMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_BlockReceived);
+    System::PacketBufferHandle receivedBlock = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !receivedBlock.IsNull());
+    NL_TEST_ASSERT(inSuite, outEvent.blockdata.Data != nullptr);
+}
+
+void SendAndVerifyBlockAck(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                           TransferSession::OutputEvent & outEvent)
+{
+    // Verify PrepareBlockAck() outputs message to send
+    CHIP_ERROR err = receiver.PrepareBlockAck();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle ackToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !ackToSend.IsNull());
+
+    // Pass BlockAck to sender and verify it was received
+    err = sender.HandleMessageReceived(ackToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_AckReceived);
+}
+
+// Test a full transfer using a responding receiver and an initiating sender, receiver drive.
+void TestInitiatingReceiverReceiverDrive(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TransferSession::OutputEvent outEvent;
+    TransferSession initiatingReceiver;
+    TransferSession respondingSender;
+
+    // Chosen arbitrarily for this test
+    uint32_t numBlockSends     = 4;
+    uint16_t transferBlockSize = 10;
+
+    uint32_t numBlocksSent = 0;
+
+    // Initialize struct with TransferInit parameters
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlagsRaw = kControl_ReceiverDrive;
+    initOptions.MaxBlockSize        = transferBlockSize;
+    char testFileDes[9]             = { "test.txt" };
+    initOptions.FileDesLength       = 9;
```

Comment:
This is incorrect usage. The file designator length should be 8 (e.g. strlen(testFileDes)), as the protocol does not expect nul-terminators.

Reply:
Got it. Should the spec make this explicit? There's no mention of null-terminators.

Reply:
If it's length-prefixed, it's never nul-terminated. There is no mention of nul-terminated strings. In network protocols, you can never expect the "nul terminator assumption" of C/C++ as it's actually not a good technique on wire formats, and not used in any common network protocol.